mock/py/mockbuild/util.py

```diff
@@ -1175,35 +1175,71 @@ def check_config(config_opts):
                                     "- option config_opts['root'] must be present in your config.")
 
 
-@traceLog()
-def include(config_file, config_opts, is_statement=False):
-    if os.path.exists(config_file):
-        if is_statement and config_file in config_opts['config_paths']:
-            getLog().warning("Multiple inclusion of %s, skipping" % config_file)
-            return
-
-        config_opts['config_paths'].append(config_file)
-
-        with open(config_file) as f:
-            content = f.read()
-            content = re.sub(r'include\((.*)\)', r'include(\g<1>, config_opts, True)', content)
-            code = compile(content, config_file, 'exec')
-        # pylint: disable=exec-used
+def _check_unrendered_config_opts(config_opts, content):
+    env = jinja2.Environment(undefined=jinja2.StrictUndefined, loader=jinja2.FileSystemLoader('/'))
+    template = env.from_string(content)
+
+    try:
+        template.render(config=config_opts)
+    except jinja2.exceptions.UndefinedError as ex:
+        m_ar = ex.message.split('attribute')
+        if len(m_ar) == 2:
+            new_msg = "Missing config option %s" % m_ar[-1].strip()
+        else:
+            new_msg = ex.message
+        raise exception.ConfigError(new_msg)
+
+
+def _render_config_template(content, config_opts):
+
+    def _non_empty_finalizer(v):
+        return None if isinstance(v, jinja2.Undefined) else v
+
+    env = jinja2.Environment(finalize=_non_empty_finalizer, loader=jinja2.FileSystemLoader('/'))
+    template = env.from_string(content)
+    part_old = None
+    while True:
+        part = template.render(config=config_opts)
+        code = compile(part, '<string>', 'exec')
         exec(code)
-    else:
-        raise exception.ConfigError("Could not find included config file: %s" % config_file)
+        if part_old == part:
+            break
```

Comment:
Here should be same safety stop. I.e. break the loop after e.g., 1000 loops anyway. Ideally, this would be controlled by some config option.

Reply:
This 'while True' loop is for the nested level references:

`config_opts['var1'] = 'value1'`
`config_opts['var2'] = {{ config_opts.var1  }}`
`config_opts['var3'] = {{ config_opts.var2  }}`

The loop will try to expand everything until nothing else is not expanded more

For code above 3 render loops:
- one for {{ config_opts.var1  }}
- second for {{ config_opts.var2  }}`
- third to get same rendering result and to stop

So, if you think there will be 1000 nested level references , consider to add such an option. But I really doubt of it

Reply:
What will happen if:
```
config_opts['var1'] = '{{ config_opts.var2 }}'
config_opts['var2'] = '{{ config_opts.var1 }}'
```
?

Reply:
This currently evaluates to:

`config_opts['var1'] = 'None'`
`config_opts['var2'] = 'None'`

Reply:
well, I can rewrite this to show something like:

mockbuild.exception.ConfigError: Missing config option 'var2'

Reply:
Uch, we did this (re-expanding templates) in distgen tool, and I'm still not convinced that calculating fixed point with jinja templates was a good step forward.  If that was the
right way of thinking, jinja would implement it itself.

Reply:
>  jinja would implement it itself

jinja2 would never implement this because this is completely irrelevant to this mock's PR, which target here is to setup correct config with feature improvements ( jinja2 templating ) rather then simple rendering templates what jinja2 does. If you read and understood the PR (and mock's ) code, you would never write such a comment

Reply:
You underestimate what I say, jinja is not macro language, but templating language.  That's what I'm saying.

Reply:
Or in other words, you implement macro language with jinja, and that's clear misconception.

Reply:
> and that's clear misconception

why?

Reply:
Because jinja language is not designed to be a macro language (tokens, AST interpolation, etc.).  It is designed for 1-hit templating.... Did you consider hiring macropy?

Reply:
> Because jinja language is not designed to be a macro language (tokens, AST interpolation, etc.). It is designed for 1-hit templating

That's you vision and I did not argue with this.

But what we create for mock here using any tools (including jinja2) can consider as a little macro language, can't it? If something in the code world can't do something, then a programmer takes this as a part, extends with other code parts and gets what he want. Easy

Reply:
Just for reference, we touched some pros/cons with this approach here https://github.com/devexp-db/distgen/pull/41